### DISCUSSION TITLE
shelving comments and patch them at the end if needed

### DISCUSSION THREAD

**Discussion by iwusong** [vuejs/rfcs#452] [RFC Discussions]
- Start Date: 2022-06-22
- Target Major Version:  3.x 
 - Reference Issues:  singleroot with comments is compiled as fragment on compilerOptions.comments = true
https://github.com/vuejs/core/pull/6026  https://github.com/vuejs/core/issues/5203 https://github.com/vuejs/core/issues/6080  https://github.com/vuejs/core/issues/5546 https://github.com/vuejs/core/pull/5832
 

# Summary
We should remove comments node in ast or transform,  record them on render vnode or instance to add them in patch if needed.

# Motivation
Comments should not affect PatchFlag and fall-through attribute inheritance in theory, it caused a lot of problems. and it will have more problems in app.config.compilerOptions.comments, dev_root_fragment, keep-alive, transition...
Even as long as we add a function, we must consider these.
Comments do not need to participate in diff or props update In fact，we can just patch them at the end if needed. This will solve the problem once and for all.

# Detailed design
 As said above
 Remove comments node in ast or transform,  record them on render vnode or instance to add them in patch if needed
# Drawbacks
There may be involve multiple parts
 

# Adoption strategy

It is not  a breaking change.

**Comment by leopiccionia** [comment]
I can see that a component with one root tag  + one or more comments becoming a fragment is confusing.

However, I don't see how silently ignoring comments in certain situations (even if `compilerOptions.comment` is explicitly set to `true`) is less confusing.

I think a better solution is:
1. Observing `compilerOptions.comments` during development/testing (not only in production);
2. Warning during development (if it doesn't already happen) when some built-in components like `Transition` or `KeepAlive` receives a fragment as children. Ideally, the warning message should have an actionable;
3. Warning during development if a component with fragment receives attrs not listed in props/emits (and `inheritAttrs` is not set to `false`).

I don't know if adding a comment is an idiom for forcing a template to be compiled as a fragment (for whatever reason), but changing it could be a breaking change.

**Reply by iwusong** [reply]
I think comment should not cause change of attribute inheritance behavior

**Reply by iwusong** [reply]
Comments may be deleted or added at any time，but PatchFlag shouldn't change

**Comment by theoephraim** [comment]
Ran into this recently and it was VERY confusing and took a long time to figure out. Please let's fix this!

Also see a related idea for being more explicit about preserving comments here - https://github.com/vuejs/rfcs/discussions/469
